File: minheap.cpp
#include <vector>			// it might be better to use a deque?
#include <iostream>
#include <cstdint>
#include <stdio.h>

template<typename ValueType>
class MinHeap
{
	std::vector<ValueType>	m_values;

public:
	typedef ValueType value_type;
	typedef MinHeap<ValueType> self_type;

	static size_t parentOfNode(size_t node) { return (node - 1) >> 1; }
	static size_t leftChildOfNode(size_t node) { return (node << 1) + 1; }

	size_t size() const { return m_values.size(); }
	bool empty() const { return m_values.empty(); }
	void clear() { m_values.clear(); }

	const value_type& front() const { return m_values[0]; }

private:
	void heap_down(size_t node)
	{
		const size_t heapSize = size();
		for (;;) {
			// find which node has the lowest value of node, leftChild and rightChild.
			size_t swapNode = node;
			// check the left child.
			size_t child = leftChildOfNode(node);
			if (child >= heapSize)
				return;
			if (m_values[child] < m_values[swapNode])
				swapNode = child;
			// check the right child.
			++child;
			if (child < heapSize && m_values[child] < m_values[swapNode])
				swapNode = child;
			else if (swapNode == node)
				return;
			std::swap(m_values[node], m_values[swapNode]);
			node = swapNode;
		}
	}

	void heap_up(size_t node)
	{
		while (node > 0) {
			size_t parentNode = parentOfNode(node);
			value_type& parentValue = m_values[parentNode];
			value_type& nodeValue = m_values[node];
			if (nodeValue < parentValue) {
				std::swap(parentValue, nodeValue);
				node = parentNode;
			} else {
				return;
			}
		}
	}

public:
	void pop_front()
	{
		if (empty())
			return;
		m_values[0] = m_values.back();
		m_values.pop_back();
		heap_down(0);
	}

	void push_back(const value_type& value)
	{
		const size_t nodePos = m_values.size();
		m_values.push_back(value);
		heap_up(nodePos);
	}
	
	void replace_front(const value_type& value)
	{
		if (size() == 0)
			push_back(value);
		else {
			m_values[0] = value;
			heap_down(0);
		}
	}
	
	bool check_heap() const
	{
		const size_t heapSize = size();
		for (size_t node = 1; node < heapSize; ++node) {
			// check that node is not < parent.
			if (m_values[node] < m_values[parentOfNode(node)])
				return false;
		}
		return true;
	}

	void dump_heap()
	{
		const size_t heapSize = size();
		size_t node = 0;
		for (size_t level = 0; node < heapSize; ++level) {
			std::cout << "|" << level << "| ";
			for (size_t col = 0; col < (1 << level); ++col) {
				if (node < heapSize)
					std::cout << m_values[node] << " ";
				else
					std::cout << "---" << " ";
				++node;
			}
		}
		std::cout << std::endl;
	}

private:
public:
	void sort()
	{
		const size_t heapSize = size();
		bool again;
		do {
			again = false;
			for (size_t node = 2; node < heapSize; ++node) {
				if (m_values[node] < m_values[node - 1]) {
					std::swap(m_values[node - 1], m_values[node]);
					heap_up(node - 1);
					again = true;
				}
			}
		} while (again);
	}
};

int main() {
	MinHeap<uint64_t> heap;
	heap.push_back(3);
	heap.push_back(2);
	heap.push_back(1);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.push_back(10);
	heap.push_back(20);
	heap.push_back(30);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.clear();
	heap.push_back(10);
	heap.push_back(20);
	heap.push_back(30);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}
	
	heap.push_back(1);
	heap.push_back(2);
	heap.push_back(3);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.clear();
	heap.push_back(100);
	heap.push_back(200);
	heap.push_back(300);
	heap.push_back(30);
	heap.push_back(10);
	heap.push_back(20);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.pop_front();
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.push_back(8);
	heap.push_back(7);
	heap.push_back(6);
	heap.push_back(3);
	heap.push_back(4);
	heap.push_back(5);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.replace_front(1);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}
	
	heap.replace_front(1000);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}
	
	for (size_t i = 0; i < 8; ++i)
		heap.push_back(8);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.replace_front(400);
	heap.push_back(9);
	heap.push_back(1);
	heap.push_back(1);
	heap.push_back(2);
	heap.push_back(9);
	heap.push_back(1005);
	heap.push_back(1010);
	heap.push_back(1008);
	heap.push_back(1004);
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}

	heap.sort();
	heap.dump_heap();
	if (!heap.check_heap()) {
		std::cout << "fail" << std::endl;
		return 0;
	}
	
}

File: minheap.py
class MinHeap(object):

    def __init__(self):
        self.array = []
        self.__add = self.array.append
        self.__pop = self.array.pop

    def empty(self):
        return len(self.array) == 0

    def size(self):
        return len(self.array)

    def front(self):
        return self.array[0]

    def push_back(self, value):
        self.__add(value)
        sz = len(self.array)
        if sz > 1:
            self.heap_up(sz-1)

    def pop_front(self):
        array = self.array
        sz = len(array)
        if sz == 0:
            return None
        if sz == 1:
            return self.__pop
        poppedValue = array[0]
        array[0] = self.__pop
        self.heap_down(0)
        return poppedValue

    def replace_front(self, value):
        if len(self.array) == 0:
            self.push_back(value)
            return
        self.array[0] = value
        self.heap_down(0)

    def heap_down(self, node):
        array = self.array
        swapNode = node
        swapValue = array[node]
        size = len(array)
        child = (node * 2) + 1
        if child < size and array[child] < swapValue:
            swapNode = child
            swapValue = array[child]
        child += 1
        if child < size and array[child] < swapValue:
            swapNode = child
            swapValue = array[child]
        if swapNode is node:
            return
        array[swapNode] = array[node]
        array[node] = swapValue
        self.heap_down(swapNode)

    def heap_up(self, node):
        if node == 0:
            return
        array = self.array
        nodeValue = array[node]
        parentNode = (node - 1) >> 1
        parentValue = array[parentNode]
        if nodeValue < parentValue:
            array[parentNode] = nodeValue
            array[node] = parentValue
            self.heap_up(parentNode)
 
    def check_heap(self):
        array = self.array
        size = len(array)
        for node in range(1, size):
            nodeValue = array[node]
            if nodeValue < array[(node - 1) >> 1]:
                print("node %u !(%u < %u)" % node)
                return False
            child = (node << 1) + 1
            if child >= size:
                continue
            if array[child] < nodeValue:
                return False
            child += 1
            if child >= size:
                continue
            if array[child] < nodeValue:
                return False
        return True

    def dump_heap(self):
        size = len(self.array)
        if size == 0:
            print("heap is empty")
            return
        depth = len(bin(size)) - 2
        print("Heap contains %u elements, so %u levels" % (size, depth))

        nodeFormat = ' %2X '
        maxNodes = (1 << (depth - 1))
        nodeFormatWidth = len(nodeFormat % 0)
        graphWidth = maxNodes * nodeFormatWidth

        node = 0
        for d in range(0, depth):
            nodesOnLine = 1 << d
            padWidth = int(((graphWidth / nodesOnLine) - nodeFormatWidth) / 2)
            padding = ' ' * padWidth
            str = "%02u: " % (d)
            for n in range(0, nodesOnLine):
                value = (nodeFormat % self.array[node]) if node < size else " -- "
                str += padding + value + padding
                node += 1
            print(str)

